R/sheets_speedread.R

```diff
@@ -0,0 +1,221 @@
+#' sheets_speedread(ss, .url = "datasource", .method = "readr")
+#'
+#' if (FALSE) {
+#' library(bench)
+#' bnch <- bench::mark(
+#'   sheets_speedread(ss, .url = "export", .method = "vroom"),
+#'   sheets_speedread(ss, .url = "export", .method = "readr"),
+#'   sheets_speedread(ss, .url = "datasource", .method = "vroom"),
+#'   sheets_speedread(ss, .url = "datasource", .method = "readr"),
+#'   iterations = 5,
+#'   check = FALSE
+#' )
+#' bnch
+#' }
+#'
+#' # prove that we can send a cell range, sheet, and col spec through ...
+#' (df <- sheets_speedread(
+#'   sheets_example("deaths"),
+#'   sheet = "other",
+#'   range = "A5:F15",
+#'   col_types = vroom::cols(
+#'     Age = vroom::col_integer(),
+#'     `Date of birth` = vroom::col_date("%m/%d/%Y"),
+#'     `Date of death` = vroom::col_date("%m/%d/%Y")
+#'   ),
+#'   .method = "vroom"
+#' ))
+#'
+#'
+#' if (FALSE) {
+#' # write a private Sheet
+#' (ss <- sheets_write(iris))
+#'
+#' # read via 'export' URL (sends a bearer token)
+#' sheets_speedread(ss, .url = "export")
+#'
+#' # read via 'datasource' URL (access token goes as query param)
+#' sheets_speedread(ss, .url = "datasource")
+#'
+#' # clean up
+#' googledrive::drive_trash(ss)
+#' }
+sheets_speedread <- function(ss,
+                             sheet = NULL,
+                             range = NULL,
+                             skip = 0,
+                             ...,
+                             .url = c("export", "datasource"),
+                             .method = c("vroom", "readr")) {
+  .method <- match.arg(.method)
+  stop_for_pkg <- function(pkg) {
+    if (!requireNamespace(pkg, quietly = TRUE)) {
+      stop_glue("The {pkg} package must be installed to use {bt('sheets_speedread()'}")
+    }
+  }
+  stop_for_pkg(.method)
+
+  ssid <- as_sheets_id(ss)
+  maybe_sheet(sheet)
+  check_range(range)
+  check_non_negative_integer(skip)
+  .url <- match.arg(.url)
+
+   x <- sheets_get(ssid)
+
+  params <- list(spreadsheet_id = unclass(ssid))
+  sheet_msg <- ""
+  range_msg <- ""
+  range_spec <- as_range_spec(
+    range, sheet = sheet, skip = skip,
+    sheets_df = x$sheets, nr_df = x$named_ranges
+  )
+  if (!is.null(range_spec$named_range)) {
+    stop_glue("{bt('sheets_speedread()'} cannot work with a named range")
+  }
+  if (!is.null(range_spec$cell_limits)) {
+    range_spec$cell_range <- as_sheets_range(range_spec$cell_limits)
+  }
+  if (!is.null(range_spec$cell_range)) {
+    params <- c(params, range = range_spec$cell_range)
+    range_msg <- glue(", range {dq(range_spec$cell_range)}")
+  }
+  if (!is.null(range_spec$sheet_name)) {
+    s <- lookup_sheet(range_spec$sheet_name, sheets_df = x$sheets)
+    params <- c(params, gid = s$id)
+    sheet_msg <- glue(", {dq(range_spec$sheet_name)} sheet")
+  }
+  message_glue("Reading from {dq(x$name)}{sheet_msg}{range_msg}")
+
+  token <- sheets_token()
+
+  if (.url == "export") {
+    params[["format"]] <- "csv"
+    params[["req_path"]] <- "export"
+  }
```

Comment:
I think this should be an `else if`, and I wonder if ` .url <- match.arg(.url)` should be immediately above it, so it's more easier to tell that `.url` only has two valid values.